src/useRouteQuery/useRouteQuery.ts

```diff
@@ -1,90 +1,111 @@
 import { reactive, watch } from 'vue'
-import { LocationQuery, LocationQueryValue, useRoute, useRouter } from 'vue-router'
+import { LocationQuery, LocationQueryValue, RouteLocationNormalizedLoaded, Router, useRoute, useRouter } from 'vue-router'
 
-type UseRouteQuery = {
+export type UseRouteQuery = {
   query: LocationQuery,
   clear: () => void,
-  set: (key: string, value: string) => void,
+  set: (key: string, value: LocationQueryValue | LocationQueryValue[]) => void,
   get: (key: string) => LocationQueryValue | LocationQueryValue[],
   remove: (key: string) => void,
 }
 
 type QueryOperation = (query: LocationQuery) => LocationQuery
 
-export function useRouteQuery(): UseRouteQuery {
-  const route = useRoute()
-  const router = useRouter()
+function factory(): () => UseRouteQuery {
   const operations = reactive<QueryOperation[]>([])
+  let route: RouteLocationNormalizedLoaded
+  let router: Router
   let interval: ReturnType<typeof setTimeout>
 
-  const clear = (): void => {
-    router.push({ query: {} })
-  }
-
-  const set = (key: string, value: string): void => {
-    query[key] = value
-  }
-
-  const get = (key: string): LocationQueryValue | LocationQueryValue[] => {
-    return query[key]
-  }
-
-  const remove = (key: string): void => {
-    delete query[key]
-  }
-
-  const query: LocationQuery = new Proxy({}, {
-    get(target, property) {
-      if (typeof property === 'symbol') {
-        return Reflect.get(target, property)
-      }
-
-      const query = applyQueryOperations(route.query, operations)
-
-      return query[property]
-    },
-    set(target, property, value) {
-      if (typeof property === 'symbol' || typeof value !== 'string') {
-        return Reflect.set(target, property, value)
-      }
-
-      const operation = updateOperationFactory(property, value)
-
-      operations.push(operation)
-
-      return true
-    },
-    deleteProperty(target, property) {
-      if (typeof property === 'symbol') {
-        return Reflect.deleteProperty(target, property)
-      }
-
-      const operation = deleteOperationFactory(property)
-
-      operations.push(operation)
-
-      return true
-    },
-  })
-
   const update = (): void => {
     const query = applyQueryOperations(route.query, operations)
 
+    operations.splice(0)
+
     router.push({ query })
   }
 
   watch(operations, () => {
+    if (!operations.length) {
+      return
+    }
+
     clearInterval(interval)
     interval = setTimeout(update)
   })
 
-  return {
-    clear,
-    set,
-    get,
-    remove,
-    query,
+  const useRouteQuery = (): UseRouteQuery => {
+    router = useRouter()
+    route = useRoute()
+
+    const clear = (): void => {
+      router.push({ query: {} })
```

Comment:
I'm surprised to see this method pushing to the router instead of resetting the `query` variable and relying on its  setter

Reply:
hmmm, I guess I could do that. But the `query` variable is just a proxy wrapper around `route.query` and the `operations` queue. And objects don't have a specific way of emptying them so it wasn't something I thought to support on the `query` proxy itself. 

Updating the url directly seems like the simplest thing to do here IMO.

Reply:
Yeah fair enough 🤷🏻